### PR TITLE
[RS Backup] PS backup support for new features

### PR DESCRIPTION
**Pull request by dragonfly91**
1. Added backup support for Azure SQL 
2. Added capability to specify retention of recovery point during backup
3. Added backup support for Encrypted VMs

### CODE REVIEW

File: src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/ScenarioTests/AzureSql/AzureSqlPolicyTests.cs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+using Microsoft.Azure.Test;
+using RecoveryServices.Tests.Helpers;
+using Microsoft.Azure;
+
+namespace RecoveryServices.Tests
+{
+    public class AzureSqlPolicyTests : RecoveryServicesTestsBase
+    {
+        [Fact]
+        public void ListRecoveryServicesAzureSqlProtectionPolicyTest()
```
Comment (markcowl) [inline src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/ScenarioTests/AzureSql/AzureSqlPolicyTests.cs]: Can we shorten this test name?  It impacts the path length of the recording file, which could cause problems depending on where you put your clone of the repo.

File: src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/ScenarioTests/AzureSql/AzureSqlPolicyTests.cs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                    Assert.NotNull(resource.Id);
+                    Assert.NotNull(resource.Name);
+                    Assert.NotNull(resource.Type);
+                    Assert.NotNull(resource.Properties);
+                }
+            }
+        }
+
+        [Fact]
+        public void GetAddUpdateDeleteAzureSqlPolicyTest()
```
Comment (markcowl) [inline src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/ScenarioTests/AzureSql/AzureSqlPolicyTests.cs]: GetAddUpdateDele=Crud, shorter name

File: src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/ScenarioTests/AzureSql/PolicyTests.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+using Microsoft.Azure.Test;
+using RecoveryServices.Tests.Helpers;
+using Microsoft.Azure;
+
+namespace RecoveryServices.Tests
+{
+    public class AzureSqlPolicyTests : RecoveryServicesTestsBase
+    {
+        [Fact]
+        public void ListRecoveryServicesProtectionPolicyTest()
```
Comment (markcowl) [inline src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/ScenarioTests/AzureSql/PolicyTests.cs]: name length

File: src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/ScenarioTests/AzureSql/PolicyTests.cs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+                    Assert.NotNull(resource.Id);
+                    Assert.NotNull(resource.Name);
+                    Assert.NotNull(resource.Type);
+                    Assert.NotNull(resource.Properties);
+                }
+            }
+        }
+
+        [Fact]
+        public void GetAddUpdateDeleteIaaSVMPolicyTest()
```
Comment (markcowl) [inline src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/ScenarioTests/AzureSql/PolicyTests.cs]: Crud

File: src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/app.config [hunk 5]
```diff
@@ -42,7 +42,20 @@
     <add key="StorageAccountId" value="/subscriptions/f2edfd5d-5496-4683-b94f-b3588c579009/resourceGroups/shraccanrg/providers/Microsoft.Storage/storageAccounts/shraccanseastorv2" />
     <add key="SourceVmIdForRestore" value="/subscriptions/f2edfd5d-5496-4683-b94f-b3588c579009/resourceGroups/shraccanrg/providers/Microsoft.Compute/virtualMachines/shraccanv2" />
 
-  </appSettings>
+   <!-- AzureSql related -->
+    <add key="WorkloadTypeAzureSqlDb" value="AzureSqlDb" />
```
Comment (markcowl) [inline src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/app.config]: This is a really bad idea - it makes it very difficult to run the tests.  The recommendation is  to generate these values inside the test setup

File: src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/app.config [hunk 6]
```diff
@@ -42,7 +42,20 @@
     <add key="StorageAccountId" value="/subscriptions/f2edfd5d-5496-4683-b94f-b3588c579009/resourceGroups/shraccanrg/providers/Microsoft.Storage/storageAccounts/shraccanseastorv2" />
```
Comment (markcowl) [inline src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/app.config]: These should beremoved.  Others running these tests will not have access to this account.  Again., it is better to do account setup in the test setup.

**Comment by markcowl** [comment]
@dragonfly91  Please make the test method name changes and change the names of the recording files accordingly.

The .config changes do not need to gate this check-in.  However, we will; require this for the next update to this library, so please start moving from reliance on existing resources to creating those resources in test setup.

**Comment by dragonfly91** [comment]
I've shortened the test names. We have a tracking task to update our tests to use automated setup. Will take it up in the next iteration. Please review / merge this PR.

**Comment by markcowl** [comment]
@azuresdkci retest this please

File: src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/Helpers/BackupTestHelper.cs [hunk 7]
```diff
@@ -38,14 +38,20 @@ public BackupTestHelper(RecoveryServicesBackupManagementClient client)
 
         public void BackupProtectedItem()
         {
-            string rsVaultRgName = "pstestrg";
-            string rsVaultName = "pstestrsvault";
+            string rsVaultRgName = "labRG1";
```
Comment (markcowl) [inline src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/Helpers/BackupTestHelper.cs]: These should really be generated at runtime and stored in the json recording

File: src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/ScenarioTests/AzureIaasVMTests/RestoreTests.cs [hunk 8]
```diff
@@ -71,6 +71,14 @@ public void RestoreDiskTest()
                     TargetDomainNameId = string.Empty,
                     TargetResourceGroupId = string.Empty,
                     TargetVirtualMachineId = string.Empty,
+                    EncryptionDetails = new EncryptionDetails()
+                    {
+                        EncryptionEnabled = true,
+                        KekUrl = ConfigurationManager.AppSettings["KekUrl"],
```
Comment (markcowl) [inline src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/ScenarioTests/AzureIaasVMTests/RestoreTests.cs]: As discussed, these values should no longer be coming from app.config - these should be created as part of the test

File: src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/app.config [hunk 9]
```diff
@@ -25,24 +25,41 @@
...
 	<!-- Restore related -->
-    <add key="RsVaultRgNameRestore" value="shankcanaryrg" />
-    <add key="RsVaultNameRestore" value="shankcanary2" />
-    <add key="RsVaultIaasVMContainerUniqueNameRestore" value="iaasvmcontainerv2;shraccanrg;shraccanv2" />
-    <add key="RsVaultIaasVMItemUniqueNameRestore" value="iaasvmcontainerv2;shraccanrg;shraccanv2" />
-    <add key="vaultLocation" value="southeastasia" />
-    <add key="RecoveryPointName" value="30641628955971" />
-    <add key="StorageAccountId" value="/subscriptions/f2edfd5d-5496-4683-b94f-b3588c579009/resourceGroups/shraccanrg/providers/Microsoft.Storage/storageAccounts/shraccanseastorv2" />
-    <add key="SourceVmIdForRestore" value="/subscriptions/f2edfd5d-5496-4683-b94f-b3588c579009/resourceGroups/shraccanrg/providers/Microsoft.Compute/virtualMachines/shraccanv2" />
+    <add key="RsVaultRgNameRestore" value="labRG1" />
```
Comment (markcowl) [inline src/ResourceManagement/RecoveryServices.Backup/RecoveryServices.Tests/app.config]: See previous comments. we need to remove this file and create any necessary resources as part of test setup

**Comment by markcowl** [comment]
@dragonfly91 some issues where we are adding to app.config - we need to make setup of necessary resources part of test setup

**Comment by dragonfly91** [comment]
@markcowl yes, we are planning to change the tests - we have a tracking task to update our tests to use automated setup which will be taken up up in the next iteration. Can this PR be merged?

**Comment by swatimadhukar** [comment]
@markcowl For the test cases related to azure sql backup it would be difficult to follow "we need to make setup of necessary resources part of test setup" at this point of time. For the Azure SQL feature cmdlets, half of the dependent cmdlets are being authored by AzureSql team(create commands) and half are created by Backup team (which is mostly get commands). Unless we setup our resource using Rest Api calls before running the test cases we cannot run these cmdlets test cases. As both the teams will be shipping respective cmdlets in the upcoming release, we can only change these test cases to run end to end only after both the cmdlets are available. We can surely work on changing the test cases in next release cycle when both teams cmdlets are available.

**Comment by markcowl** [comment]
@swatimadhukar We will require this in the next release, along with the conversion to AutoRest

**Comment by swatimadhukar** [comment]
@markcowl We will surely make all the required changes in the tests in next release when both team's SQL and Backup cmdlets for Azure Sql backup are released and is ready for use.

**Comment by dragonfly91** [comment]
@markcowl Hi, the tests have passed; can you continue with the review / merge?

**Pull request merged**